File: python/ghleak/scanner.py
"""TruffleHog wrapper — runs scans with verification enabled."""

from __future__ import annotations

import json
import logging
import subprocess
import tempfile
from pathlib import Path
from typing import Any

from . import config

log = logging.getLogger(__name__)


class TruffleHogResult:
    """Wraps a single verified finding from TruffleHog."""

    def __init__(self, raw: dict[str, Any]) -> None:
        self.raw = raw
        self.source_metadata: dict[str, Any] = raw.get("SourceMetadata", {})
        data = raw.get("Data", {})
        self.detector = raw.get("DetectorName", "")
        self.detector_type = raw.get("DetectorType", 0)
        self.decoder = raw.get("DecoderName", "")
        self.verified = raw.get("Verified", False)
        self.redacted = raw.get("Redacted", False)
        self.raw_v2 = data.get("RawV2", "")

        meta = self.source_metadata.get("Data", {}).get("Git", {})
        self.commit = meta.get("commit", "")
        self.file = meta.get("file", "")
        self.email = meta.get("email", "")
        self.timestamp = meta.get("timestamp", "")
        self.repository = meta.get("repository", "")
        self.committer = meta.get("committer", "")

    def __repr__(self) -> str:
        v = "VERIFIED" if self.verified else "unverified"
        return f"[{v}] {self.detector} in {self.file} @ {self.commit[:12]}"


class TruffleHogScanner:
    """Runs TruffleHog on git repos, diffs, or raw text."""

    def __init__(
        self,
        binary: str = "trufflehog",
        only_verified: bool = True,
        fail_on_error: bool = False,
        extra_args: list[str] | None = None,
    ):
        self.binary = binary
        self.only_verified = only_verified
        self.fail_on_error = fail_on_error
        self.extra_args = extra_args or []

    def scan_git_path(self, path: Path) -> list[TruffleHogResult]:
        """Scan a local git repository with TruffleHog."""
        cmd = [
            self.binary, "git", "file://.",
            "--since-commit", "HEAD",
            "--json",
        ]
        if self.only_verified:
            cmd.append("--only-verified")
        cmd.extend(self.extra_args)

        results = self._run(cmd, cwd=path)
        log.info("Scanned %s: %d findings, %d verified",
                 path, len(results), sum(1 for r in results if r.verified))
        return results

    def scan_diff(self, diff_text: str) -> list[TruffleHogResult]:
        """Scan a plain-text diff with TruffleHog."""
        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".diff", delete=False, prefix="ghleak_"
        ) as f:
            f.write(diff_text)
            diff_path = f.name

        cmd = [
            self.binary, "filesystem", diff_path,
            "--json",
        ]
        if self.only_verified:
            cmd.append("--only-verified")
        cmd.extend(self.extra_args)

        results = self._run(cmd)
        Path(diff_path).unlink(missing_ok=True)
        return results

    def scan_commit_from_gh(
        self, owner: str, repo: str, sha: str, github_token: str | None = None
    ) -> list[TruffleHogResult]:
        """Scan a GitHub commit directly (clones + scans)."""
        clone_url = f"https://github.com/{owner}/{repo}.git"
        if github_token:
            clone_url = f"https://x-access-token:{github_token}@github.com/{owner}/{repo}.git"

        tmp = Path(tempfile.mkdtemp(prefix="ghleak_"))
        try:
            subprocess.run(
                ["git", "clone", "--depth", "1", "--single-branch", clone_url, str(tmp)],
                capture_output=True, timeout=120, check=True,
            )
            subprocess.run(
                ["git", "-C", str(tmp), "fetch", "--depth", "1", "origin", sha],
                capture_output=True, timeout=30, check=True,
            )
            subprocess.run(
                ["git", "-C", str(tmp), "checkout", sha],
                capture_output=True, timeout=30, check=True,
            )
            return self.scan_git_path(tmp)
        finally:
            import shutil
            shutil.rmtree(tmp, ignore_errors=True)

    def _run(self, cmd: list[str], cwd: str | Path | None = None) -> list[TruffleHogResult]:
        """Execute trufflehog and parse JSON output."""
        log.debug("Running: %s", " ".join(cmd))
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=300,
                cwd=str(cwd) if cwd else None,
            )
        except subprocess.TimeoutExpired:
            log.warning("TruffleHog timed out: %s", " ".join(cmd[:4]))
            return []
        except FileNotFoundError:
            log.error("TruffleHog binary not found at %s", self.binary)
            raise

        results: list[TruffleHogResult] = []
        for line in proc.stdout.strip().split("\n"):
            line = line.strip()
            if not line:
                continue
            try:
                raw = json.loads(line)
                results.append(TruffleHogResult(raw))
            except json.JSONDecodeError:
                continue

        if self.fail_on_error and proc.returncode not in (0, 1):
            log.error("TruffleHog failed (exit %d): %s", proc.returncode, proc.stderr[:500])

        return results


def summarize_findings(results: list[TruffleHogResult]) -> dict[str, Any]:
    """Aggregate scan results into a summary dict."""
    detector_counts: dict[str, int] = {}
    verified_count = 0
    file_paths: set[str] = set()
    emails: set[str] = set()
    repos: set[str] = set()

    for r in results:
        detector_counts[r.detector] = detector_counts.get(r.detector, 0) + 1
        if r.verified:
            verified_count += 1
        if r.file:
            file_paths.add(r.file)
        if r.email:
            emails.add(r.email)
        if r.repository:
            repos.add(r.repository)

    return {
        "total_findings": len(results),
        "verified_count": verified_count,
        "detectors": dict(sorted(detector_counts.items(), key=lambda x: -x[1])),
        "unique_files": len(file_paths),
        "unique_emails": len(emails),
        "unique_repos": list(repos),
    }

File: go/internal/scanner.go
package internal

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"sort"
	"strings"
	"sync"

	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
	"github.com/trufflesecurity/trufflehog/v3/pkg/engine"
	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/source_metadatapb"
	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/sourcespb"
	"github.com/trufflesecurity/trufflehog/v3/pkg/sources"
)

// TruffleHogScanner wraps trufflehog for secret verification.
// Uses native Go engine when possible; subprocess CLI as fallback.
type TruffleHogScanner struct {
	OnlyVerified bool
	UseEngine    bool
	Binary       string
	FailOnError  bool
	ExtraArgs    []string
}

func NewTruffleHogScanner(onlyVerified bool) *TruffleHogScanner {
	return &TruffleHogScanner{OnlyVerified: onlyVerified, UseEngine: true, Binary: "trufflehog"}
}

// ScanDiff runs trufflehog against a raw diff string.
func (s *TruffleHogScanner) ScanDiff(diff string) []Finding {
	if diff == "" {
		return nil
	}

	if s.UseEngine {
		findings, err := s.scanEngine(diff)
		if err == nil {
			return findings
		}
		log.Printf("Engine unavailable (%v), falling back to subprocess", err)
	}
	return s.scanSubprocess(diff)
}

// scanGitPath scans a local git repository path with trufflehog git subcommand.
func (s *TruffleHogScanner) scanGitPath(path string) []Finding {
	if s.UseEngine {
		findings, err := s.scanGitEngine(path)
		if err == nil {
			return findings
		}
		log.Printf("Git engine unavailable (%v), falling back to subprocess", err)
	}

	cmd := exec.Command(s.Binary,
		"git", "file://.",
		"--since-commit", "HEAD",
		"--json", "--no-update",
	)
	if s.OnlyVerified {
		cmd.Args = append(cmd.Args, "--only-verified")
	}
	cmd.Args = append(cmd.Args, s.ExtraArgs...)
	cmd.Dir = path

	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	output, err := cmd.Output()
	if err != nil && len(output) == 0 {
		if s.FailOnError {
			log.Fatalf("trufflehog git failed: %v\nstderr: %s", err, stderr.String())
		}
		log.Printf("trufflehog git stderr: %s", stderr.String())
		return nil
	}
	return parseTHJSON(string(output))
}

// scanCommitFromGH clones a GitHub repo (depth=1), checks out a specific SHA,
// and scans the checkout with trufflehog filesystem.
func (s *TruffleHogScanner) scanCommitFromGH(owner, repo, sha, tmpDir, token string) []Finding {
	cloneURL := fmt.Sprintf("https://github.com/%s/%s.git", owner, repo)
	if token != "" {
		cloneURL = fmt.Sprintf("https://x-access-token:%s@github.com/%s/%s.git", token, owner, repo)
	}

	if err := os.MkdirAll(tmpDir, 0755); err != nil {
		log.Printf("failed to create tmp dir %s: %v", tmpDir, err)
		return nil
	}

	cloneCmd := exec.Command("git", "clone", "--depth", "1", "--single-branch", cloneURL, tmpDir)
	if out, err := cloneCmd.CombinedOutput(); err != nil {
		log.Printf("git clone failed: %v\n%s", err, string(out))
		return nil
	}

	fetchCmd := exec.Command("git", "-C", tmpDir, "fetch", "--depth", "1", "origin", sha)
	if out, err := fetchCmd.CombinedOutput(); err != nil {
		log.Printf("git fetch sha failed: %v\n%s", err, string(out))
		return nil
	}

	checkoutCmd := exec.Command("git", "-C", tmpDir, "checkout", sha)
	if out, err := checkoutCmd.CombinedOutput(); err != nil {
		log.Printf("git checkout failed: %v\n%s", err, string(out))
		return nil
	}

	return s.scanGitPath(tmpDir)
}

// summarizeFindings aggregates findings into a summary.
func summarizeFindings(findings []Finding) map[string]any {
	detectorCounts := map[string]int{}
	verifiedCount := 0
	fileSet := map[string]struct{}{}
	emailSet := map[string]struct{}{}
	repoSet := map[string]struct{}{}

	for _, f := range findings {
		detectorCounts[f.DetectorName]++
		if f.Verified {
			verifiedCount++
		}
		if f.File != "" {
			fileSet[f.File] = struct{}{}
		}
		if f.Email != "" {
			emailSet[f.Email] = struct{}{}
		}
		if f.Repository != "" {
			repoSet[f.Repository] = struct{}{}
		}
	}

	uniqueRepos := make([]string, 0, len(repoSet))
	for r := range repoSet {
		uniqueRepos = append(uniqueRepos, r)
	}
	sort.Strings(uniqueRepos)

	return map[string]any{
		"total_findings": len(findings),
		"verified_count": verifiedCount,
		"detectors":      detectorCounts,
		"unique_files":   len(fileSet),
		"unique_emails":  len(emailSet),
		"unique_repos":   uniqueRepos,
	}
}

// ── Native engine path ─────────────────────────────────────────────

func (s *TruffleHogScanner) scanEngine(diff string) ([]Finding, error) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	resultsCh := make(chan detectors.ResultWithMetadata, 200)

	eng, err := engine.NewEngine(ctx, &engine.Config{
		Concurrency:   4,
		Verify:        s.OnlyVerified,
		SourceManager: sources.NewManager(),
		Dispatcher:    &dispatcherCollector{ch: resultsCh},
	})
	if err != nil {
		return nil, fmt.Errorf("engine init: %w", err)
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		eng.Start(ctx)
	}()

	eng.ScanChunk(&sources.Chunk{
		Data:       []byte(diff),
		SourceType: sourcespb.SourceType_SOURCE_TYPE_FILESYSTEM,
		SourceName: "ghleak",
		SourceMetadata: &source_metadatapb.MetaData{
			Data: &source_metadatapb.MetaData_Filesystem{
				Filesystem: &source_metadatapb.Filesystem{File: "diff"},
			},
		},
		SourceVerify: s.OnlyVerified,
	})

	if err := eng.Finish(ctx); err != nil {
		log.Printf("engine finish err: %v", err)
	}
	close(resultsCh)
	wg.Wait()

	var findings []Finding
	for r := range resultsCh {
		f := Finding{
			DetectorName: r.DetectorName,
			DetectorType: r.DetectorType.String(),
			Decoder:      r.DecoderType.String(),
			Verified:     r.Verified,
			Redacted:     r.Redacted != "",
			RawV2:        string(r.RawV2),
		}
		if meta := r.SourceMetadata.GetFilesystem(); meta != nil {
			f.File = meta.GetFile()
		}
		if meta := r.SourceMetadata.GetGit(); meta != nil {
			f.File = meta.GetFile()
			f.Commit = meta.GetCommit()
			f.Email = meta.GetEmail()
			f.Timestamp = meta.GetTimestamp()
			f.Repository = meta.GetRepository()
		}
		findings = append(findings, f)
	}
	return findings, nil
}

type dispatcherCollector struct {
	ch chan detectors.ResultWithMetadata
}

func (d *dispatcherCollector) Dispatch(ctx context.Context, result detectors.ResultWithMetadata) error {
	d.ch <- result
	return nil
}

// scanGitEngine uses the native Go engine to scan a local git repository.
func (s *TruffleHogScanner) scanGitEngine(path string) ([]Finding, error) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	resultsCh := make(chan detectors.ResultWithMetadata, 200)

	eng, err := engine.NewEngine(ctx, &engine.Config{
		Concurrency:   4,
		Verify:        s.OnlyVerified,
		SourceManager: sources.NewManager(),
		Dispatcher:    &dispatcherCollector{ch: resultsCh},
	})
	if err != nil {
		return nil, fmt.Errorf("engine init: %w", err)
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		eng.Start(ctx)
	}()

	if _, err := eng.ScanGit(ctx, sources.GitConfig{
		URI: fmt.Sprintf("file://.%s%s", string(os.PathSeparator), path),
	}); err != nil {
		return nil, fmt.Errorf("engine git scan: %w", err)
	}

	if err := eng.Finish(ctx); err != nil {
		log.Printf("engine finish err: %v", err)
	}
	close(resultsCh)
	wg.Wait()

	var findings []Finding
	for r := range resultsCh {
		f := Finding{
			DetectorName: r.DetectorName,
			DetectorType: r.DetectorType.String(),
			Decoder:      r.DecoderType.String(),
			Verified:     r.Verified,
			Redacted:     r.Redacted != "",
			RawV2:        string(r.RawV2),
		}
		if meta := r.SourceMetadata.GetFilesystem(); meta != nil {
			f.File = meta.GetFile()
		}
		if meta := r.SourceMetadata.GetGit(); meta != nil {
			f.File = meta.GetFile()
			f.Commit = meta.GetCommit()
			f.Email = meta.GetEmail()
			f.Timestamp = meta.GetTimestamp()
			f.Repository = meta.GetRepository()
		}
		findings = append(findings, f)
	}
	return findings, nil
}

// ── Subprocess path ────────────────────────────────────────────────

func (s *TruffleHogScanner) scanSubprocess(diff string) []Finding {
	cmd := exec.Command(s.Binary,
		"filesystem", "--no-update", "--json",
	)
	if s.OnlyVerified {
		cmd.Args = append(cmd.Args, "--only-verified")
	}
	cmd.Args = append(cmd.Args, s.ExtraArgs...)
	cmd.Args = append(cmd.Args, "-")
	cmd.Stdin = bytes.NewReader([]byte(diff))

	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	output, err := cmd.Output()
	if err != nil && len(output) == 0 {
		if s.FailOnError {
			log.Fatalf("trufflehog subprocess failed: %v\nstderr: %s", err, stderr.String())
		}
		log.Printf("trufflehog stderr: %s", stderr.String())
		return nil
	}
	return parseTHJSON(string(output))
}

func parseTHJSON(raw string) []Finding {
	var findings []Finding
	scanner := bufio.NewScanner(strings.NewReader(raw))
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var data map[string]any
		if err := json.Unmarshal([]byte(line), &data); err != nil {
			continue
		}

		f := Finding{}

		if v, ok := data["DetectorName"].(string); ok {
			f.DetectorName = v
		}
		f.DetectorType = fmt.Sprintf("%v", data["DetectorType"])
		if v, ok := data["DecoderName"].(string); ok {
			f.Decoder = v
		}
		if v, ok := data["Verified"].(bool); ok {
			f.Verified = v
		}
		if v, ok := data["Redacted"].(bool); ok {
			f.Redacted = v
		}

		// RawV2 from Data.RawV2 or top-level RawV2
		if rd, ok := data["Data"].(map[string]any); ok {
			if rv2, ok := rd["RawV2"].(string); ok {
				f.RawV2 = rv2
			}
		} else if rv2, ok := data["RawV2"].(string); ok {
			f.RawV2 = rv2
		}

		// SourceMetadata — support both with and without Data wrapper
		if sm, ok := data["SourceMetadata"].(map[string]any); ok {
			var gitData, fsData map[string]any
			if d, ok := sm["Data"].(map[string]any); ok {
				gitData, _ = d["git"].(map[string]any)
				fsData, _ = d["filesystem"].(map[string]any)
			}
			if gitData == nil {
				gitData, _ = sm["git"].(map[string]any)
			}
			if fsData == nil {
				fsData, _ = sm["filesystem"].(map[string]any)
			}

			if fsData != nil {
				if file, ok := fsData["file"].(string); ok {
					f.File = file
				}
			}
			if gitData != nil {
				if file, ok := gitData["file"].(string); ok {
					f.File = file
				}
				if commit, ok := gitData["commit"].(string); ok {
					f.Commit = commit
				}
				if email, ok := gitData["email"].(string); ok {
					f.Email = email
				}
				if ts, ok := gitData["timestamp"].(string); ok {
					f.Timestamp = ts
				}
				if repo, ok := gitData["repository"].(string); ok {
					f.Repository = repo
				}
				if committer, ok := gitData["committer"].(string); ok {
					f.Committer = committer
				}
			}
		}

		findings = append(findings, f)
	}
	return findings
}
